server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -112,42 +132,77 @@ private void maybeScheduleNextRun(TimeValue time) {
+            && HealthMetadata.getFromClusterState(event.state()) != null
+            && HealthNode.findHealthNode(event.state()) != null;
+        if (prerequisitesFulfilled) {
+            DiscoveryNode previous = HealthNode.findHealthNode(event.previousState());
+            DiscoveryNode current = HealthNode.findHealthNode(event.state());
+            if (Objects.equals(previous, current) == false) {
+                // The new health node (probably) does not have any information yet, so the last
+                // reported health info gets reset to null.
+                lastReportedDiskHealthInfo.set(null);
+            }
         }
+        maybeScheduleNow();
     }
 
-    // Visible for testing
-    void monitorHealth() {
-        if (inProgress.compareAndSet(false, true)) {
+    private void monitorHealth() {
+        if (inProgress.compareAndSet(false, true) && prerequisitesFulfilled) {
             ClusterState clusterState = clusterService.state();
             HealthMetadata healthMetadata = HealthMetadata.getFromClusterState(clusterState);
-            assert healthMetadata != null : "health metadata should have been initialized.";
-            DiskHealthInfo previousHealth = this.lastReportedDiskHealthInfo;
+            if (healthMetadata == null) {
+                logger.debug("Couldn't retrieve health metadata.");
+                inProgress.set(false);
+                return;
+            }
+            DiskHealthInfo previousHealth = this.lastReportedDiskHealthInfo.get();
             DiskHealthInfo currentHealth = diskCheck.getHealth(healthMetadata, clusterState);
             if (currentHealth.equals(previousHealth) == false) {
-                logger.debug("Health status changed from {} to {}", previousHealth, currentHealth);
-                this.lastReportedDiskHealthInfo = currentHealth;
+                String nodeId = clusterService.localNode().getId();
+                ActionListener<AcknowledgedResponse> listener = ActionListener.wrap(response -> {
+                    // Update the last reported value if there was no change, that would mean that there
+                    // is a new health node that might have not received the value, so we should retry.
+                    lastReportedDiskHealthInfo.compareAndSet(previousHealth, currentHealth);
+                    logger.debug(
+                        "Health info [{}] successfully send, last reported value: {}.",
+                        currentHealth,
+                        lastReportedDiskHealthInfo.get()
+                    );
+                }, e -> logger.error(() -> format("Failed to send health info [%s] to health node, will try again.", currentHealth), e));
+                client.execute(
+                    UpdateHealthInfoCacheAction.INSTANCE,
+                    new UpdateHealthInfoCacheAction.Request(nodeId, currentHealth),
+                    ActionListener.runAfter(listener, () -> {
```

Comment:
I know we're FWind the request, but we know who the health node is. 

Should we not direct the request to the specific node here? (avoid - most of the times - some network hops)

Reply:
Is this a network hop? I thought it would resolve if it's local or send it to the health node directly? If I understand you correctly, that assumption is wrong?

Reply:
Ah you're correct it does execute locally (and `TransportHealthNodeAction` does the FW if necessary).

Nevermind :) Sorry for the noise (I was under the impression a random node is chosen)

Reply:
No worries, it's good to verify these things ;).